Allow dotted hosts creation using config

Aravindh-Raju opened the pull request:
Changes in this pull request:
- Added a config with settings which are used to allow the creation of dotted hosts.
- Details about the config can be found in `modules/docs/src/main/mdoc/operator/config-api.md`.

Jay07GIT commented:
On creating the dotted hosts A record in the zones, I could see weird error message Can you please update this.

Steps to reproduce: 
1. Create dummy. zone.
2. Create dotted hots with the record `ok.ok.` ->  record name, record type ->  A, Ip Address -> 9.5.6.7

<img width="1292" alt="image" src="https://user-images.githubusercontent.com/88772914/193509897-8167fd43-9b22-46da-ad4a-d56b22c1d069.png">

Aravindh-Raju commented:
> Can you please update this message. <img alt="image" width="413" src="https://user-images.githubusercontent.com/88772914/193506900-7124f29e-0f27-46a3-be1a-6fbaaf1af7e0.png">

Updated the message, only indicates that the record is a dotted host now and removed the other contents.

Aravindh-Raju commented:
> On creating the dotted hosts A record in the zones, I could see weird error message Can you please update this.
> 
> Steps to reproduce:
> 
> 1. Create dummy. zone.
> 2. Create dotted hots with the record `ok.ok.` ->  record name, record type ->  A, Ip Address -> 9.5.6.7
> 
> <img alt="image" width="1292" src="https://user-images.githubusercontent.com/88772914/193509897-8167fd43-9b22-46da-ad4a-d56b22c1d069.png">

Thanks for finding that out. It was because of the dot at the end of the record name. Made the required changes. Works as expected now.

nspadaccino reviewed (commented):
Couple more things I noticed:
• Currently there seems to be no limit on how many dots could be in a dotted host. For example I could create a record "foo.foo.foo.foo.foo." in zone "ok." Depending on requirements, maybe we could have this be a configurable value
• In the case of wildcard zones, we could have conflicts in the access rules. For example, if we have dotted host config with zone = "*ent.com." and config with zone = "child.parent.com.", which would take precedence for zone "child.parent.com." ? Does the order of the config matter?
• It seems like if allowed-group-list and allowed-user-list are both left empty, any user can create a dotted record in the zone. Is this what we want, or should we allow nobody if those fields are both blank? 
• It looks like if allowed-record-type is left empty, users can only create A records, while other record types fail, which doesn't seem like the intended behavior

Aravindh-Raju commented on modules/api/src/main/scala/vinyldns/api/domain/record/RecordSetValidations.scala:
```diff
@@ -90,6 +90,68 @@ object RecordSetValidations {
+  ): Either[Throwable, Unit] =
+    ensuring(
+      InvalidRequest(
+        s"Record with fqdn '${newRecordSet.name}.${zone.name}' cannot be created. " +
```
Steps to create this exception:
1. Add dotted hosts config for the zone as `"*ent.com"`.
2. Create two zones `child.parent.com` and `parent.com`. Create a dotted host record in `child.parent.com` zone for (eg) record name = `test.dotted`. Then create a new dotted host record in `parent.com` zone with the same name `test.dotted.child`. This will throw the above error as the record with same type and FQDN already exist in `child.parent.com` zone.

nspadaccino replied:
Thanks, makes sense. I was able to reproduce the error

Aravindh-Raju commented on modules/api/src/test/scala/vinyldns/api/domain/record/RecordSetValidationsSpec.scala:
```diff
@@ -184,34 +186,82 @@ class RecordSetValidationsSpec
+        leftValue(isDotted(test, okZone, None, true, false)) shouldBe an[InvalidRequest]
+      }
+
+      "return a failure for a dotted record name that matches the zone name" in {
```
I should’ve removed that test case. We should allow creation of records with record name same as the zone name. We already do that before these changes which we can see when we run locally. There is no validation not to be the same. So, I think we can allow that as before. I'll remove the test case as it wasn’t necessary and added accidentally .

Aravindh-Raju commented on modules/docs/src/main/mdoc/operator/config-api.md:
```diff
@@ -536,7 +536,57 @@ v6-discovery-nibble-boundaries {
+
+For eg, we can't create a dotted host with `CNAME` record type in the zone `dummy.` as it's not in `allowed-record-type`.
+And the user `professor` can't create a dotted host in the zone `dummy.` as the user is not in `allowed-user-list` or 
+`allowed-group-list`.
```
Yes. I've made the changes in the latest commit. Added about not being part of `dummy-group`. Please have a look.

Aravindh-Raju commented on modules/api/src/test/functional/tests/recordsets/create_recordset_test.py:
```diff
@@ -524,8 +526,35 @@ def test_create_dotted_a_record_not_apex_fails(shared_zone_test_context):
+                           "zone '" + zone_name + "'"))
+
+
+def test_create_dotted_a_record_succeeds_if_all_dotted_hosts_config_satisfies(shared_zone_test_context):
```
Made the change.

Aravindh-Raju commented on modules/api/src/test/functional/tests/recordsets/create_recordset_test.py:
```diff
@@ -581,22 +610,52 @@ def test_create_dotted_a_record_apex_with_trailing_dot_succeeds(shared_zone_test
+                           "zone '" + zone["name"] + "'"))
+
+
+def test_create_dotted_cname_record_succeeds_if_all_dotted_hosts_config_satisfies(shared_zone_test_context):
```
Made the change.

Aravindh-Raju commented on modules/api/src/main/scala/vinyldns/api/domain/record/RecordSetService.scala:
```diff
@@ -169,6 +195,136 @@ class RecordSetService(
+    }
+  }
+
+  // Check if user is allowed to create dotted hosts using the users present in dotted hosts config
```
Made the change.

Aravindh-Raju commented:
> Couple more things I noticed:
> 
> • Currently there seems to be no limit on how many dots could be in a dotted host. For example I could create a record "foo.foo.foo.foo.foo." in zone "ok." Depending on requirements, maybe we could have this be a configurable value
> 
> • In the case of wildcard zones, we could have conflicts in the access rules. For example, if we have dotted host config with zone = "*ent.com." and config with zone = "child.parent.com.", which would take precedence for zone "child.parent.com." ? Does the order of the config matter?
> 
> • It seems like if allowed-group-list and allowed-user-list are both left empty, any user can create a dotted record in the zone. Is this what we want, or should we allow nobody if those fields are both blank?
> 
> • It looks like if allowed-record-type is left empty, users can only create A records, while other record types fail, which doesn't seem like the intended behavior

1. Sure, I'll add a config for number of dots allowed for each zones in the `dotted-host-config` and update the PR.
2. No, the order of the config doesn’t matter. In the end, we group all the zones together (both wildcard and individual) and process them. So, the order doesn't matter. But I noted that, currently if a zone is present in both `wildcard` and as `individual` in the `dotted-host-config`, it combines both of their properties instead of just taking the individual zone in precedence. Will look into it.
3. No, it doesn't allow any user to create a dotted record in the zone if `allowed-group-list` and `allowed-user-list` are left empty. I checked it once again today and if they're left empty, it didn’t allow anyone to create a dotted host record in that zone, which is the intended behaviour. Did it allow any user to create dotted host record for you if left empty? Can you please try again once more with latest commit?
4. Same as above. If `allowed-record-type` is left empty, user cannot create dotted host record of any type. It works fine when I ran locally and I'm able to create record of other types without any problem. Not sure what the problem is.

deepakmohanakrishnan07 commented on modules/api/src/main/scala/vinyldns/api/config/DottedHostsConfig.scala:
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2018 Comcast Cable Communications Management, LLC
```
Shall we change the copyright to have 2022 year in it?

Aravindh-Raju replied:
Contacted the open source team as suggested. Will make the required changes once we hear from them.

nspadaccino replied:
I think we can update the copyrights in another PR so we can get this merged

nspadaccino commented on modules/docs/src/main/mdoc/operator/config-api.md:
```diff
@@ -551,6 +551,7 @@ Note the following:
 5. If the user is either in `allowed-user-list` or `allowed-group-list`, they are allowed to create a dotted host. It is
 not necessary for the user to be in both `allowed-user-list` and `allowed-group-list`.
 6. The record types which are allowed while creating a dotted host is added to the `allowed-record-type`.
+7. The number of dots allowed in a record name for a zone is given in `allowed-dots-limit`.
```
I think we should specify here that the number of allowed dots is set to zero (i.e. no dotted hosts allowed) if a value is not provided. We should also specify what happens if the other config values are not provided

Aravindh-Raju replied:
Updated the doc with those details. Please have a look at it.

nspadaccino commented:
> 3. No, it doesn't allow any user to create a dotted record in the zone if `allowed-group-list` and `allowed-user-list` are left empty. I checked it once again today and if they're left empty, it didn’t allow anyone to create a dotted host record in that zone, which is the intended behaviour. Did it allow any user to create dotted host record for you if left empty? Can you please try again once more with latest commit?

I tried this again and it works as you said. I probably missed something in my test config

nspadaccino commented:
> 4. Same as above. If `allowed-record-type` is left empty, user cannot create dotted host record of any type. It works fine when I ran locally and I'm able to create record of other types without any problem. Not sure what the problem is.
 
Tried this again and it works as you said.

Jay07GIT commented on modules/api/src/main/scala/vinyldns/api/domain/record/RecordSetValidations.scala:
```diff
@@ -278,6 +357,29 @@ object RecordSetValidations {
+  def checkAllowedDots(allowedDotsLimit: Int, recordSet: RecordSet, zone: Zone): Either[Throwable, Unit] = {
+    ensuring(
+      InvalidRequest(
+        s"RecordSet with name ${recordSet.name} has more dots than that is allowed in config for this zone " +
```
This Message looks good for dots limit greater than 1. If the dots limits is set to 0, `Record with name "" and type "" is a dotted host which is not allowed in zone ""` could be valid error message.

Jay07GIT commented on modules/api/src/main/resources/reference.conf:
```diff
@@ -90,6 +90,29 @@ vinyldns {
+      {
+      # for wildcard zones. Settings will be applied to all matching zones
+      zone = "dummy*."
+      allowed-user-list = ["sharedZoneUser"]
```
A minor change can be do.

In `allowed-user-list` allowed is already mentioned above in config. So the key name can be `user-list`. What do you think?

If it's fine, we can do the similar changes for other key words in dotted hosts config as well.

Aravindh-Raju replied:
Made the changes as suggested. Please have a look.

nspadaccino reviewed (approved):
Looks good to me!
